src/Microsoft.AspNet.Server.Kestrel/Networking/SockAddr.cs

```diff
@@ -0,0 +1,94 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Net;
+
+namespace Microsoft.AspNet.Server.Kestrel.Networking
+{
+    public struct SockAddr
+    {
+        // this type represents native memory occupied by sockaddr struct
+        // https://msdn.microsoft.com/en-us/library/windows/desktop/ms740496(v=vs.85).aspx
+        // although the c/c++ header defines it as a 2-byte short followed by a 14-byte array,
+        // the simplest way to reserve the same size in c# is with four nameless long values
+        private long _field0;
+        private long _field1;
+        private long _field2;
+        private long _field3;
+
+        public SockAddr(long ignored)
+        {
+            _field3 = _field0 = _field1 = _field2 = _field3 = 0;
+        }
+
+        public IPEndPoint GetIPEndPoint()
+        {
+            // The bytes are represented in network byte order.
+            //
+            // Example 1: [2001:4898:e0:391:b9ef:1124:9d3e:a354]:39179
+            //
+            // 0000 0000 0b99 0017  => The third and fourth bytes 990B is the actual port
+            // 9103 e000 9848 0120  => IPv6 address is represented in the 128bit field1 and field2.
+            // 54a3 3e9d 2411 efb9     Read these two 64-bit long from right to left byte by byte.
+            // 0000 0000 0000 0000
+            //
+            // Example 2: 10.135.34.141:39178 when adopt dual-stack sockets, IPv4 is mapped to IPv6
+            //
+            // 0000 0000 0a99 0017  => The port representation are the same
+            // 0000 0000 0000 0000
+            // 8d22 870a ffff 0000  => IPv4 occupies the last 32 bit: 0A.87.22.8d is the actual address.
+            // 0000 0000 0000 0000
+            //
+            // Example 3: 10.135.34.141:12804, not dual-stack sockets
+            // 8d22 870a fd31 0002  => sa_family == AF_INET (02)
+            // 0000 0000 0000 0000
+            // 0000 0000 0000 0000
+            // 0000 0000 0000 0000
+
+
+            // Quick calculate the port by mask the field and locate the byte 3 and byte 4
+            // and then shift them to correct place to form a int.
+            var port = ((int)(_field0 & 0x00FF0000) >> 8) | (int)((_field0 & 0xFF000000) >> 24);
+
+            var family = (int)_field0 & 0x000000FF;
+            if (family == 0x00000002)
+            {
+                // AF_INET => IPv4
+                return new IPEndPoint(new IPAddress((_field0 >> 32) & 0xFFFFFFFF), port);
+            }
+            else if (IsIPv4MappedToIPv6())
+            {
+                var ipv4bits = (_field2 >> 32) & 0x00000000FFFFFFFF;
```

Comment:
Why not just get rid of the mask (and the other one as well)? It's a no-op after all.
Or `uint.MaxValue`